support conditional encryption

ericHao22 opened the pull request:
## Implementation

### Backend

1. add method `shouldBeEncryptable` on `Securable` trait and it would return **true** value by default
2. **(Additional)** do not generate key when available encryption key is not existed in key manager
3. **(Additional)** create a command for key creation
4. **(Additional)** fix issue of creating relationship with encryption key model in saving event of securable model

ericHao22 committed:
chore(concern): [Securable] add method to determine if a model should be encrytable

ericHao22 committed:
refactor: [KeyManager] rename method name, argument and variable

ericHao22 committed:
chore: [KeyManager] change to use provider name as type of encryption key

ericHao22 committed:
refactor(contract): [KeyProvider] remove unused method

ericHao22 committed:
chore: [KeyManager] do not generate key when available encryption key is not existed

ericHao22 committed:
chore(concern): [Securable] should throw exception when encrypting data without available key

ericHao22 committed:
chore(command): [GenerateKey] new

ericHao22 committed:
docs: [readme] update usage information

ericHao22 committed:
chore(config): [composer] execute php insights under all directories

ericHao22 committed:
chore(config): [phpinsights] adjust minimum requirement of style metric

ericHao22 committed:
chore(command): [RotateKey] check whether current key is existed before rotating

FishChen1202 commented on src/Concerns/Securable.php:
```diff
@@ -39,8 +39,17 @@ public function isEncrypted(): bool
         return (bool) $this->encryptionKeys->first();
     }
 
+    public function shouldBeEncryptable(): bool
```
不太理解這個 trait 的用法
這個 trait 主要應該是讓 use 他的 model 可以指定哪些欄位要加密 ?
假如某個 model 選擇使用 use Securable 應該就代表他都要作加密？
會有 model 掛這個 trait 但他不要加密嗎

ericHao22 replied:
主要想要保留彈性，畢竟實際上使用外部服務進行加密也會產生成本，所以營運上應該是可以針對同一張table內非機敏的資料去省略加密的流程

但如果未來發現還是不合理的話，也是會狠心把他deprecate掉的

FishChen1202 replied:
但我以為省略的部分是倚靠 他會在 model 提供需要加密的欄位才作加密 其他欄位就不去做加密？

ericHao22 replied:
我的想法是在確定哪些欄位要加密的大前提是究竟這一筆資料有沒有需要做加密？如果不需要的話也就不用再去看欄位是否需要加密了，所以這個方法目的就是拿來判斷這點

FishChen1202 commented on src/Concerns/Securable.php:
```diff
@@ -39,8 +39,17 @@ public function isEncrypted(): bool
         return (bool) $this->encryptionKeys->first();
```
假如他可以選擇要不要加密的話 用有沒有 key 來判斷是否加密過 會準確嗎？

ericHao22 replied:
這個方法主要是拿來判斷該筆資料是否曾經加密過，如果他當初就選擇不加密的話，那判斷結果也自然會是**false**

FishChen1202 replied:
Table 的一筆資料 每次加密 就會產生一組相對應的 key 這樣嗎

ericHao22 replied:
準確一點講是會關聯到一把data key，而多筆資料可能會用到同一把這樣

FishChen1202 reviewed (commented):
comments

ericHao22 committed:
fix(concern): [Securable] save data immediately without model events when encryptoin done

ericHao22 committed:
fix(observer): [ModelObserver] change to encrypt data when model saved
there might create relationship with encryption key model when calling encrypt method.
thus, we must execute encryption when a model had alreday created
